src/Microsoft.DotNet.Interactive.CSharp/CSharpKernel.cs

```diff
@@ -298,9 +299,9 @@ public async Task HandleAsync(SubmitCode submitCode, KernelInvocationContext con
             if (_currentDirectory != currentDirectory)
             {
```

Comment:
I can't get my head around the fact that this code exists in the code base (knowing my luck I could even have added it :-)).  A different thread can call SetWorkingDirectory between 298 and 300, I am assuming this is intended to maintain some sort of invariant that the cache and resolver use the current working directory.

Anything that changes "the current working directory for the notebook needs to be independent of Directory.GetWorkingDirectory()/Directory.SetWorkingDirectory().  A command to change the working directory seems like the right way to go, and that command need not ... probably (definitely in my head) should not invoke Directory.SetWorkingDirectory( ... ).

I don't believe there is any invariant that we can maintain around Directory.GetWorkingDirectory() returning us anything "notebook usefull".  It should ideally be reserved for script authors and perhaps interoperability with third party stuff that relies on it.

Reply:
I think we need a separate concept which is the notebook directory, which is not mutable at runtime.

Reply:
I'm okay with it being mutable, so long as the mutability is controlled.  The problem with Directory.GetCurrentDirectory() is that there are a bunch of ways of changing it, and we only see the effects.  So a command to change the Project Root or whatever we decide to call it is fine, probably even necessary for example when an IDE implements the "save as" gesture.

Reply:
We have a command for this currently: [ChangeWorkingDirectory](https://github.com/dotnet/interactive/blob/8438829658e70de5ff20a172748cc10f8654d1da/src/Microsoft.DotNet.Interactive/Commands/ChangeWorkingDirectory.cs#L10-L10), which is used as you're describing by the VS Code extension. This might be the wrong name for it.

Reply:
Maybe "ScriptRootDirectory" would be a good name for it. From the perspective of the VS Code extension you would understand what it will do.
And it should probably not change the current WorkingDirectory is that correct?

Reply:
The working directory should default to this location as it does today, but it would not change when the user changes the working directory.

Reply:
I've pushed another change that handles the ChangeWorkingDirectory-Command.

In the VS Code extension it's now possible to do:  Directory.SetCurrentDirectory("..")
but the "#load"-directive is not affected by that.